doc/designs/event-queue.md

```diff
@@ -0,0 +1,181 @@
+at the time the timer is supposed to trip.  To make timer events timely,
+they should be considered very high priority.
+
+For all event related structures and functions, `ossl_event_` is the naming
+prefix.  If this becomes a public API, the prefix will be upcased as usual,
+i.e. become `OSSL_EVENT_`.
+
+### Event
+
+An event is a small short-lived structure carrying information:
+
+-   *[mandatory]* An event type, which is a simple numeric identity, the
+    meaning of which is not known by the event functionality itself.
+-   *[mandatory]* A reference to an event context
+-   *[optional]* A reference to auxilliary identifying information.
+-   *[optional]* A reference to a payload, which is data passed with the
+    event (for example, a read event would carry along a buffer with data)
+-   *[optional]* A reference to a desctructor function, if there are fields
+    that need customized destruction when the event structure itself is to
+    be destroyed.
+
+The event itself is designed for a single synchronous thread, i.e. cannot be
+shared by multiple threads.  The diverse objects it refers to may, however,
+be shared by multiple threads, at the discretion of the functions in the
+method structure.
+
+Once populated, the event type, the references to event context, and the
+reference to the destructor function are considered immutable, up until the
+event structure is destroyed.
+
+The reference to the auxilliary identifying material or to the payload,
+however, are considered mutable.  Any event handler may "steal" them and
+replace the reference to them in the event structure with NULL.  Stealing
+must be done with much care.
+
+Events may be dynamically allocated.  In that case, ownership is passed to
+the event dispatch functionality (see below), and the event will be
+automatically destroyed when not needed any more.
+
+Events may also be embedded in another structure, as a static variable, or
+as an local variable in a function.  In that case, owndership remains with
+that structure or that variable.
+
+### Event queue
+
+An event queue is a structure holding the following:
+
+-   A reference to a method structure to handle the event queue.
+-   A reference to auxilliary data for the methods.
+-   An array of subscribers, which are simple callbacks taking a reference
+    to an event structure, along with auxilliary arguments.
+-   Buckets of events, one bucket for each priority level.  The number of
+    buckets depends on how many priority levels are implemented by the
+    functions of the method structure.
+    If there are zero priority levels, there are no buckets at all, and
+    events "added" to such a queue structure are always dispatched
+    immediately, if the function used to "add" the event supports that.
+
+For each event added to the queue, there is associated data:
+
+-   *[optional]* A time value, to indicate when the event is to be fired
+    off (i.e. when it's going to be passed to subscribers)
```

Comment:
Oh....wait - now I'm confused. This seems to contradict what was said earlier: "For timers, it means that timer events are expected to be added to the queue at the time the timer is supposed to trip"